libevm/lib/service_state.go

```diff
@@ -231,17 +241,11 @@ func (s *Service) StateRemoveStorage(params StorageParams) error {
 	return nil
 }
 
-// this is basically a copy of stateObject.empty() which is not exported and therefore not accessible here
-func isAccountEmpty(statedb *state.StateDB, addr common.Address) bool {
-	obj := statedb.GetOrNewStateObject(addr)
-	return obj.Nonce() == 0 && obj.Balance().Sign() == 0 && bytes.Equal(obj.CodeHash(), emptyCodeHash)
-}
-
-// chunk keys are generated by hashing the original key and the chunk index
+// chunk keys are generated by hashing a salt, the original key and the chunk index
+// the salt was added to reduce the risk of accidental hash collisions because similar strategies
+// to generate storage keys might be used by the caller
 func getChunkKey(key common.Hash, chunkIndex int) common.Hash {
-	chunkIndexBytes := big.NewInt(int64(chunkIndex)).Bytes()
-	hashedData := append(key.Bytes(), chunkIndexBytes...)
-	return crypto.Keccak256Hash(hashedData)
+	return crypto.Keccak256Hash(chunkKeySalt.Bytes(), key.Bytes(), common.BigToHash(big.NewInt(int64(chunkIndex))).Bytes())
```

Comment:
We may keep `chunkKeySalt` already in bytes, to avoid getting bytes representation every time we use read\write chunked record.
Also for `chunkIndex` do we really need this heavy construction `common.BigToHash(big.NewInt(int64(chunkIndex))).Bytes()`? Is there any simpler solution?

Reply:
Regarding `chunkKeySalt`: This is internally already a byte array. This would only create a new go-slice, basically a view, of the array. I changed it anyway.

Regarding the int conversion:
That construction is heavy and ugly, I agree, but it does a bunch of things:
- Using a `big.Int` takes care of encoding/decoding incl. endianness and the inner `int64(chunkIndex)` is just a type cast because the constructor only accepts `int64`.
- The outermost ` common.BigToHash()` takes care of properly padding to the correct length (32 bytes).

I refactored it to make clearer what is happening.